### ISSUE TITLE
Assembler.assemble's circular dependency check is not thread-safe

### ISSUE THREAD

**Issue by mzipay** [mzipay/Aglyph#3]
The `Assembler.assemble(component_spec)` method checks for circular dependencies by maintaining a stack of the component IDs that are recursively assembled for any given call to assemble.

However, multiple threads may call assemble() and, if any of those calls might include the same component ID at any point while the "assembly stack" is pushed, then a false-negative AglyphError will be raised because Assembler will believe it has encountered a circular dependency.

To reproduce:

```xml
<component id="component-1" dotted-name="SomeClass">
  <init><arg reference="component-2" /></init>
</component>
<component id="component-2" dotted-name="AnotherClass">
  <init><arg reference="component-3" /></init>
</component>
<component id="component-3" dotted-name="SomeClass">
  <init><arg><None /></arg></init>
</component>
```

```python
threads = [
  threading.Thread(target=assembler.assemble, args=("component-1",))
  for in in range(100)]
for t in threads:
  t.start()
for t in threads:
  t.join()
```

This results in many false "circular dependency detected" AglyphErrors.

**Comment by mzipay** [comment]
Using a lock to synchronize the assembly stack won't solve this - individual threads may still attempt to assemble the same component IDs, so even if each thread has a lock on the assembly stack, a duplicate component ID in the stack would still be (incorrectly) "found."

**Comment by mzipay** [comment]
Fixed in 3.0.0 by using `threading.local()` for the assembly stack (see commit a96d27a).
